File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.cpp
typedef long long ll;
class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        vector<ll> pre(nums.size()+1);
        pre[1]=nums[0];
        for(int i=2;i<=nums.size();i++)
        {
            pre[i]+=pre[i-1]+nums[i-1];
        }
        deque<ll> dq;
        ll ans=1e9;
        for(int i=0;i<pre.size();i++)
        {
            // calculate the answer
            while(dq.size()&&pre[i]-pre[dq.front()]>=k)
            {
                ans=min(ans,i-dq.front());
                dq.pop_front();
            }
            // remove 
            while(dq.size()&&pre[i]<=pre[dq.back()])
            {
                dq.pop_back();
            }
            dq.push_back(i);
        }
        return (ans==1e9?-1:ans);
    }
};

File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.java
class Solution {
    public int shortestSubarray(int[] nums, int k) {
        ArrayDeque<Integer> dq=new ArrayDeque<>();
        int n=nums.length,ans=Integer.MAX_VALUE;
        int res=n+1;
        long[] pre=new long[res];
        for(int i=1;i<n+1;i++)
        {
            pre[i]+=nums[i-1]+pre[i-1];
        }
        for(int i=0;i<pre.length;i++)
        {
            while(dq.size()>0&&pre[i]-pre[dq.getFirst()]>=k)
            {
                ans=Math.min(ans,i-dq.peekFirst());
                dq.pollFirst();
            }
            while(dq.size()>0&&pre[i]<=pre[dq.getLast()])
            {
                dq.pollLast();
            }
            dq.addLast(i);
        }
        return (ans==Integer.MAX_VALUE?-1:ans);
    }
}